src/main/java/seedu/duke/storage/FoodListDecoder.java

```diff
@@ -1,49 +1,233 @@
 
+        if (encodedFood.equals("")) {
+            throw new EmptyStorageFileException();
+        }
+        String[] details = encodedFood.split(SEPARATOR);
+        String foodName = "";
+        String expiryDate = "";
+        String quantity = "0.0";
+        String unit = "";
+        String category = "";
+
+
+        for (String foodDetails : details) {
+            try {
+                if (foodDetails.equals("")) {
+                    continue;
+                }
+
+                char identifier = foodDetails.charAt(0);
+                String detail = foodDetails.substring(1).trim();
+
+                switch(identifier) {
+                case NAME_IDENTIFIER:
+                    if (hasValidFoodName(detail)) {
+                        assert !detail.equals("");
+                        foodName = detail;
+                    }
+                    break;
+                case EXPIRY_DATE_IDENTIFIER:
+                    if (hasValidExpiryDate(detail)) {
+                        expiryDate = detail;
+                    }
+                    break;
+                case QUANTITY_IDENTIFIER:
+                    if(hasValidQuantity(detail)) {
+                        quantity = detail;
+                    }
+                    break;
+                case UNIT_IDENTIFIER:
+                    if (hasValidUnit(detail)) {
+                        unit = detail;
+                    } else {
+                        unit = String.valueOf(Unit.UNIT);
+                    }
+                    break;
+                case CATEGORY_IDENTIFIER:
+                    if(hasValidCategory(detail)) {
+                        category = detail;
+                    } else {
+                        category = String.valueOf(FoodCategory.UNCLASSIFIED_FOOD);
+                    }
+                    break;
+                default:
+                    break;
+                }
+            } catch (StringIndexOutOfBoundsException e) {
+                Ui.showError(MISSING_IDENTIFIER_ERROR);
+            }
+        }
         try {
-            return new Food(details[0].trim(), details[1].trim()
-                    , parseDouble(details[2].trim()), details[3].trim(), details[4].trim());
-        } catch (ArrayIndexOutOfBoundsException e) {
-            throw new StorageOperationException("Unable to read file successfully");
+            return new Food(foodName, expiryDate, Double.valueOf(quantity), unit, category);
+        } catch (NumberFormatException e) {
+            Ui.showError(INVALID_QUANTITY_ERROR);
         }
+        return new Food(foodName, expiryDate, category);
     }
+
+    /**
+     * This method checks whether the encoded <code>foodName</code> is valid.
+     *
+     * @param detail is the encoded foodName
+     * @return true if valid
+     * @throws ReadStorageFileErrorException when not valid
+     */
+    private static boolean hasValidFoodName(String detail) throws ReadStorageFileErrorException {

```

Comment:
Why do you implement the validation here if there is already a generic `Validator` class in the utils? Please take a look on it cause this basically repeats the same code.

Reply:
Yes, but the methods are slightly different. For example, the Boolean methods I wrote for `unit` and `category` only return `false` rather than throwing an exception when valid. This is so as I can set the default value for these 2 attributes instead of deleting the whole line of data. I feel that a single corrupted attribute should not warrant the whole food object to be ignored; especially as `unit` and `category` are not very significant attributes (users can probably deduce the values of these 2 attributes based on `quantity` and `foodName` respectively).